### DISCUSSION TITLE
OpenOCD下载问题，求助！

### DISCUSSION THREAD

**Discussion by HangX-Ma** [ZhuYanzhen1/miniFOC#13] [Q&A]
用的nano-dap进行烧录，环境完全按照作者的配置，进行烧录的时候，执行`reset run`出现如下报错：
```shell
Error: stm32f1x.cpu -- clearing lockup after double fault
Polling target stm32f1x.cpu failed, trying to reexamine
Info : stm32f1x.cpu: hardware has 6 breakpoints, 4 watchpoints
```
网上有人说把Boot0引脚拉高一下就行，但是作者的方案里面已经是接地了，不好拉引脚。

自己折腾了一会儿，后来又出现`flash write_bank 0 xxx.hex 0`就会卡死出错的情况：
```shell
SWD DPIDR 0x1ba01477
Failed to write memory at 0x20000060
Failed to get read pointer
...
```
毫无OpenOCD以及GD32的调试经验，希望得到帮助。

**Comment by ZhuYanzhen1** [comment]
通过SWD引脚可以直接进行烧录，无需拉高BOOT0引脚。因为改变BOOT引脚唯一改变的就是上电后的引导方式，而SWD引脚（即调试器接口）是通过ARM CoreSight调试组件进行程序的下载的，与引导无关。我在google上找到了一些相似的问题，可以看看对你有没有帮助：https://forum.sparkfun.com/viewtopic.php?t=41014。也可以先检查一下芯片和焊接的问题，有一些拆机的芯片的储存器会被锁死导致无法下载程序。

**Reply by HangX-Ma** [reply]
在telnet中尝试过`stm32f1x unlock 0`指令，仍旧会出现如上描述的第一类错误，几次尝试上锁以及解锁的指令后，flash变成如上所述的第二类错误，无法读取写入。之后我试试重新焊接芯片，可能焊接时29号接地的PAD的引脚处的焊锡有溢出导致引脚短接，芯片外观上观察其他引脚焊点并无问题。

**Reply by ZhuYanzhen1** [reply]
好的，也可以尝试换一个烧录器，不同烧录器在配置上有略微区别，我使用的是ST-Link V2。还可以检查一下OpenOCD烧录时的复位选项，建议设置为none或srst。

**Reply by HangX-Ma** [reply]
很奇怪，已经不知道是什么地方出问题了，叙述一下调试过程：
- 焊接部分检测过后没问题，怀疑是不是OpenOCD程序本身有问题导致无法下载或者烧录器有问题，就找了STM32F767ZI的板子（STLink-V2-1)，因为OpenOCD的scripts本身有对stm32系列的支持，所以用的是官方给的cfg脚本。ST官方给的Keil程序能实现点灯，但用OpenOCD提示烧写成功（但我看到烧写进去的字节数永远为32768，和实际不符），可无法亮灯，反汇编之后程序卡死在某个特定地址，满屏的`movs R0, R0`。
- `reset_config none`设置也没有效果。
- 另外自查了OpenOCD的flash相关的指令，并进行检测，可以肯定的是之前有烧录进去一部分，但是地址和内容都和源文件对不上。而之后用STM32的板子测试，烧录是成功的，内容也对的上。

暂时也是没辙了，Google都快翻烂了。:ghost:

**Reply by HangX-Ma** [reply]
ST的问题已经解决了，我在写CMake的时候参考了一个博主的文章，他在写的时候有这么一句
```CMake
add_link_options(
    ...
    LINKER:--build-id
)
```
`The GNU linker`描述说明该命令选项会创建`.note.gnu.build-id`的 `.section`，用`arm-none-eabi-objdump -h xxx.elf`解析elf
文件，正确的文件的`.section`应当如下所示，而`.note.gnu.build-id`的`.section`会被插入到`Idx`为0处。我猜测可能是因爲这样导致SymstmInit初始化系统时钟的时候找不到中断向量表程序跑飞。我再测试一下之前的那块板子，如果还不行等ST-Link到了之后再进行测试，后续进展我更新到这个discussion上，方便有需要的人参考借鉴。
```shell
Sections:
Idx Name          Size      VMA       LMA       File off  Algn
  0 .isr_vector   000001f8  08000000  08000000  00010000  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  1 .text         00000bdc  080001f8  080001f8  000101f8  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  2 .rodata       00000010  08000dd4  08000dd4  00010dd4  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  3 .init_array   00000004  08000de4  08000de4  00010de4  2**2
                  CONTENTS, ALLOC, LOAD, DATA
  4 .fini_array   00000004  08000de8  08000de8  00010de8  2**2
                  CONTENTS, ALLOC, LOAD, DATA
  5 .data         00000004  20000000  08000dec  00020000  2**2
                  CONTENTS, ALLOC, LOAD, DATA
  6 .bss          00000034  20000004  08000df0  00020004  2**2
                  ALLOC
  7 ._user_heap_stack 00000600  20000038  08000df0  00020038  2**0
                  ALLOC
  8 .ARM.attributes 00000030  00000000  00000000  00020004  2**0
                  CONTENTS, READONLY
  9 .debug_info   000061c5  00000000  00000000  00020034  2**0
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 10 .debug_abbrev 0000119c  00000000  00000000  000261f9  2**0
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 11 .debug_aranges 000003e8  00000000  00000000  00027398  2**3
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 12 .debug_ranges 000003b0  00000000  00000000  00027780  2**3
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 13 .debug_line   000044ac  00000000  00000000  00027b30  2**0
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 14 .debug_str    00001dad  00000000  00000000  0002bfdc  2**0
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 15 .comment      00000049  00000000  00000000  0002dd89  2**0
                  CONTENTS, READONLY
 16 .debug_frame  00000814  00000000  00000000  0002ddd4  2**2
                  CONTENTS, READONLY, DEBUGGING, OCTETS
 17 .debug_loc    00002bf7  00000000  00000000  0002e5e8  2**0
                  CONTENTS, READONLY, DEBUGGING, OCTETS
```

**Reply by HangX-Ma** [reply]
经过排查，发现原来的芯片上电后用openocd烧录的时候一直会发生`external reset`，焊接部分检查重新焊接也没有解决问题。最后更换芯片，openocd成功烧录，且板子能够运行程序。

具体执行步骤如下：
- cmsis-dap.cfg
```tcl
# choose st-link/j-link/dap-link etc.
adapter driver cmsis-dap
transport select swd
# source [find interface/stlink.cfg]
# transport select hla_swd
source [find target/stm32f1x.cfg]

# download speed = 1MHz
adapter speed 1000
```
- 挂载板子并烧录
```shell
openocd -f cmsis-dap.cfg -c "reset halt;flash write_image erase xxx.bin 0x08000000;reset run;exit"
```
- 或者
```shell
openocd -f cmsis-dap.cfg
```
使用`MobaXterm` 指定 `hostname: localhost`，`port: 4444`，之后键入
```shell
reset halt
flash write_image erase xxx.bin 0x08000000
reset run
exit
```


作者所写的指令中`flash probe 0`是检查flash bank序号为0处的目标，可以通过`flash banks`查看具体有几个bank以及其各自的序号；另外`stm32f1x mass_erase 0`和之前命令中的erase是同样的作用；而`flash write_bank 0 xxx.hex 0`指定在bank 0中写入hex文件，offset为0。
